### PR TITLE
Added ‘Notes’ to ‘Job’ so users can record some notes about each job …

### PR DESCRIPTION
**Pull request by steven-hill**
…application.

Adding a new job application to the tracker:
	- tapping on ‘+’ in HomeViewController will present a sheet (AddNewJobViewController).
	- users can enter text for the company, position, location, choose a status from the popup button, and add notes.

Editing a previously saved job application:
	- users can edit a job application by tapping on its cell.
	- the job application details will be presented in a sheet (JobDetailViewController).
	- users can make changes and tap ‘Done’ to save them or ‘Cancel’ to discard them.

Added a status label to both sheets to indicate that the status can be chosen from the popup button beside it.

The dashboard now shows dynamic data and updates when new applications are added or when existing ones are edited.

Implemented CoreData for persistence. If there is a CoreData error, an alert is shown to the user.

Now that the data comes from CoreData, the ‘Job’ file was deleted. This file was only used for dummy data while building the UI.

Added search functionality by company name:
    - the app gets the search query from the search bar.
    - it performs a fetch request using the query string.
    - it assigns the results to jobsArray and displays the results in the collection view.
    - while the search results are shown, a button opposite the ‘Applications’ label is visible and can be tapped to allow users to go back to the main list.
    - this button is only visible to the user when search results are displayed in the collection view.
    - if the search query returns no results, an alert notifies the user.

### CODE REVIEW

File: Job Application Tracker/JobApplicationTracker/JobApplicationTracker/AppDelegate.swift [hunk 1]
```diff
@@ -9,28 +9,17 @@ import UIKit
...
 
     // MARK: UISceneSession Lifecycle
 
     func application(_ application: UIApplication, configurationForConnecting connectingSceneSession: UISceneSession, options: UIScene.ConnectionOptions) -> UISceneConfiguration {
-        // Called when a new scene session is being created.
-        // Use this method to select a configuration to create the new scene with.
         return UISceneConfiguration(name: "Default Configuration", sessionRole: connectingSceneSession.role)
     }
 
     func application(_ application: UIApplication, didDiscardSceneSessions sceneSessions: Set<UISceneSession>) {
```
Comment (OlivAlex) [inline Job Application Tracker/JobApplicationTracker/JobApplicationTracker/AppDelegate.swift]: Can we remove all these empty methods altogether? 🤔

File: Job Application Tracker/JobApplicationTracker/JobApplicationTracker/CoreData/CoreDataManager.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import Foundation
+import CoreData
+
+class CoreDataManager {
+
+    var context: NSManagedObjectContext {
+        return CoreDataManager.manager.persistentContainer.viewContext
+    }
+    
+    init() {
```
Comment (OlivAlex) [inline Job Application Tracker/JobApplicationTracker/JobApplicationTracker/CoreData/CoreDataManager.swift]: Do we need the empty init?

File: Job Application Tracker/JobApplicationTracker/JobApplicationTracker/AppDelegate.swift [hunk 3]
```diff
@@ -9,28 +9,17 @@ import UIKit
 
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
-
```
Comment (OlivAlex) [inline Job Application Tracker/JobApplicationTracker/JobApplicationTracker/AppDelegate.swift:12]: ❤️ Love the cleanup!

File: Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/AddNewJobViewController.swift [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        newJob.position = positionTextfield.text!
+        newJob.location = locationTextfield.text!
+        newJob.status = statusPopUpButton.currentTitle!
+        newJob.notes = notesTextView.text!
+        
+        delegate?.addNewJob(self, addedJob: newJob)
+        self.dismiss(animated: true, completion: nil)
+    }
+    
+    func layoutViews() {
```
Comment (OlivAlex) [inline Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/AddNewJobViewController.swift:70]: I see we're setting a lot of top, leading, trailing, height constraints. Could we maybe add the UI element to an array and loop over to set those constraints?

File: Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/HomeViewController.swift [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+                let ok = UIAlertAction(title: "Ok", style: .default) { _ in
+                    self.dismiss(animated: true, completion: nil)
+                }
+                alert.addAction(ok)
+                self.present(alert, animated: true, completion: nil)
+            }
+            return
+        }
+        
+        let queryText = searchBar.text!
```
Comment (OlivAlex) [inline Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/HomeViewController.swift]: Could we add this to the guard above so we don't rely on forcefully unwrapping?

File: Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/HomeViewController.swift [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+            } else {
+                jobsArray = results
+                DispatchQueue.main.async {
+                    self.returnToAllButton.isHidden = false
+                    self.collectionView.reloadData()
+                }
+            }
+        } catch {
+            print(error)
+            DispatchQueue.main.async {
```
Comment (OlivAlex) [inline Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/HomeViewController.swift:117]: I see we're presenting a lot of alerts, could we maybe have a func that takes in an enum of our possible errors and presents the right alert UI?

File: Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/JobDetailViewController.swift [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+//  JobDetailViewController.swift
+//  JobApplicationTracker
+//
+//  Created by Steven Hill on 22/09/2022.
+//
+
+import UIKit
+import CoreData
+
+protocol JobDetailViewControllerDelegate: AnyObject {
```
Comment (OlivAlex) [inline Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/JobDetailViewController.swift]: Do we need the AnyObject here and in the other protocols?

**Comment by OlivAlex** [comment]
LGTM! Lets just look at these comments and see if we can clean it up a bit! :)

**Comment by steven-hill** [comment]
Changes made post-PR review. 

(A new commit has been pushed to the same branch.)

Removed empty methods from AppDelegate and SceneDelegate.

Removed line spaces between the IBOutlets.

Removed the isTextEntered variable from HomeViewController and rewrote the guard statement in makeSearchQuery() to remove the forcibly unwrapped optionals.

Added the missing ‘private’ keyword to the configure() method in the PopUpButton file.

Removed ‘AnyObject’ from AddNewJobViewControllerDelegate and JobDetailViewControllerDelegate.

Removed the empty init() method in the CoreDataManager file.

Re-indented code in the three view controllers.

Updated the constraints on the storyboard for the ‘Dashboard’ and ‘Applications’ labels, and ‘ReturnToAllButton’. The previous constraints cut off some of the text in the labels on iPhone SE. Now the entire text of both labels is visible on iPhone SE.

**Review by OlivAlex — APPROVED** [review]
This looks great. Thanks for this!

**Pull request merged**